halo2_frontend/src/plonk/circuit/constraint_system.rs

```diff
@@ -0,0 +1,1241 @@
+        let mut selector_replacements = vec![None; selector_assignment.len()];
+        for assignment in selector_assignment {
+            selector_replacements[assignment.selector] = Some(assignment.expression);
+            selector_map[assignment.selector] = Some(new_columns[assignment.combination_index]);
+        }
+
+        self.selector_map = selector_map
+            .into_iter()
+            .map(|a| a.unwrap())
+            .collect::<Vec<_>>();
+        let selector_replacements = selector_replacements
+            .into_iter()
+            .map(|a| a.unwrap())
+            .collect::<Vec<_>>();
+        self.replace_selectors_with_fixed(&selector_replacements);
+        self.selectors_to_fixed = true;
+
+        (self, selectors_to_fixed)
+    }
+
+    /// This will compress selectors together depending on their provided
+    /// assignments. This `ConstraintSystem` will then be modified to add new
+    /// fixed columns (representing the actual selectors) and will return the
+    /// polynomials for those columns. Finally, an internal map is updated to
+    /// find which fixed column corresponds with a given `Selector`.
+    ///
+    /// Do not call this twice. Yes, this should be a builder pattern instead.
+    #[deprecated(note = "Use `selectors_to_fixed_compressed` instead")]
+    pub fn compress_selectors(self, selectors: Vec<Vec<bool>>) -> (Self, Vec<Vec<F>>) {
+        let (cs, selectors_to_fixed) = self.selectors_to_fixed_compressed();
+        let fixed_polys = selectors_to_fixed.convert(selectors);
+
+        (cs, fixed_polys)
+    }
+
+    /// Transform this `ConstraintSystem` into an equivalent one that replaces the selector columns
+    /// by fixed columns with a direct mapping.
+    ///
+    /// Panics if called twice.
+    pub fn selectors_to_fixed_direct(mut self) -> (Self, SelectorsToFixed) {
+        if self.selectors_to_fixed {
+            panic!("the selectors have already been transformed to fixed columns");
+        }
+        let selectors_to_fixed = SelectorsToFixed {
+            compress: false,
+            num_selectors: self.num_selectors,
+            max_degree: 0,
+            degrees: vec![],
+        };
+
+        let selector_replacements: Vec<_> = (0..self.num_selectors)
+            .map(|_| {
+                let column = self.fixed_column();
+                let rotation = Rotation::cur();
+                Expression::Fixed(FixedQuery {
+                    index: Some(self.query_fixed_index(column, rotation)),
+                    column_index: column.index,
+                    rotation,
+                })
+            })
+            .collect();
+
+        self.replace_selectors_with_fixed(&selector_replacements);
+        self.selectors_to_fixed = true;
+        (self, selectors_to_fixed)
+    }
+
+    /// Does not combine selectors and directly replaces them everywhere with fixed columns.
+    pub fn directly_convert_selectors_to_fixed(
```

Comment:
TODO: Mark as deprecated

Reply:
Done in https://github.com/privacy-scaling-explorations/halo2/pull/290/commits/15fb6278df713c71b3448ef9f2a1bcae06c70393